docs/proposals/generic-web-api-provider-config.md

```diff
@@ -0,0 +1,145 @@
+
+### Detailed examples
+
+#### Example 1: Authenticating to Upbound
+
+[Upbound][upbound] uses 60 minute short-lived tokens to authenticate users to access their managed control planes using their robot tokens (API keys). In order to exchange their robot token, users must make a POST request to the `/orgscopedtokens` API passing a form body that includes the name of the Upbound organization to authenticate against and the robot token. If successful, this API returns a JSON payload which includes a string `access_token` to be used as the bearer token when communicating with the Upbound control plane.
+
+```yaml!
+webAPIProviderConfig:
+  method: "POST"
+  url: https://auth.upbound.io/apis/tokenexchange.upbound.io/v1alpha1/orgscopedtokens
+  # this example uses ${.orgName} and {.token}, which are provided by the user as part of the secret
+  body: "audience=upbound%3Aspaces%3Aapi&audience=upbound%3Aspaces%3Acontrolplanes&grant_type=urn%3Aietf%3Aparams%3Aoauth%3Agrant-type%3Atoken-exchange&scope=upbound%3Aorg%3A${.orgName}&subject_token={.token}&subject_token_type=urn%3Aietf%3Aparams%3Aoauth%3Atoken-type%3Aid_token"
+  headers:
+    "Content-Type": "application/x-www-form-urlencoded"
+  tokenPath: $.access_token
+```
+
+[upbound]: https://www.upbound.io/
+
+### Security Considerations
+
+This proposal suggests that Argo CD make calls to arbitrary endpoints. Ideally the user would be able to validate the security of the endpoint independently before configuring Argo CD to interact with it.
+
+Argo CD will require the use of TLS in its requests to the external APIs to protect against man-in-the-middle attacks that could compromise users' credentials during transmission.
+
+Argo CD will only store the credentials returned by the APIs in memory and will never persist them to a volume or secret that may be accessible by other users.
+
+### Risks and Mitigations
+
+
+
+### Upgrade / Downgrade Strategy
+
+No action is needed by the user to upgrade to the latest version of Argo CD that supports generic web API provider configurations. The feature is entirely opt-in and users that do not choose to opt-in will see no change.
+
+When downgrading, users that have opted-in to using the configuration will not be able to access clusters that are using it to authenticate. Users will either need to revert to long-lived credentials or will need to find workaround solutions to authenticating these clusters using short-lived credentials (such as the ones mentioned in the "Motivation" section).
+
+## Drawbacks
+
+Asking users to provide the parameters for an authentication request is a poor user experience. Users may not be aware of the authentication APIs for their cluster providers and therefore might not know what values are valid for their Argo CD cluster. This poor experience may be offset by documentation, provided either by Argo and/or by the cluster providers, with examples of how to configure the requests.
+
+## Alternatives
```

Comment:
Maybe an alternative to document is a solution external to ArgoCD? A pod that lives in the argocd namespace can perform query to the cluster provider, get the token and the "interface" with Kubernetes/ArgoCD by modifying the Kubernetes secret that represent the cluster. 

This solution, if it is viable and I understood the problem correctly, is not coupled to ArgoCD and is more generic since it only uses the Kubernetes API. A simple pod deployment that is configured with a CM could be hosted on argoproj-labs. It also has the advantage of not increasing the complexity in ArgoCD.

Would this work as an alternative?

Reply:
Yeah I could see that working as well. There would still need to be some amount of modification to ArgoCD to support that new interface, correct?

But that also feels nicely generic sufficient to support any HTTP authentication